src/main/scala/ai/privado/dataflow/DuplicateFlowProcessor.scala

```diff
@@ -167,46 +167,89 @@ object DuplicateFlowProcessor {
     ruleCache: RuleCache
   ): Unit = {
 
+    val expendedSourceSinkInfo = processExpendedSourceSinkData(dataflowMapByPathId, privadoScanConfig, ruleCache)
+
+    expendedSourceSinkInfo.foreach(flow => {
+      AuditCache.addIntoBeforeSecondFiltering(SourcePathInfo(flow.pathSourceId, flow.sinkId, flow.sinkPathId))
+      filterFlowsOverlappingWithOtherDataElement(
+        flow.pathSourceId,
+        flow.dataflowsMapByType,
+        flow.dataflowSinkType,
+        flow.dataflowNodeType,
+        ruleCache,
+        flow.sinkId,
+        flow.sinkPathId
+      )
+    })
+  }
+
+  case class ExpendedSourceSinkInfo(
+    pathSourceId: String,
+    sinkId: String,
+    dataflowSinkType: String,
+    dataflowNodeType: String,
+    dataflowsMapByType: Map[String, Path],
+    sinkPathId: String
+  )
+
+  def processExpendedSourceSinkData(
+    dataflowMapByPathId: Map[String, Path],
+    privadoScanConfig: PrivadoInput,
+    ruleCache: RuleCache,
+    isStoreMatricData: Boolean = true
+  ): List[ExpendedSourceSinkInfo] = {
     val sinkSubCategories = mutable.HashMap[String, mutable.Set[String]]()
     ruleCache.getRule.sinks.foreach(sinkRule => {
       if (!sinkSubCategories.contains(sinkRule.catLevelTwo))
         sinkSubCategories.addOne(sinkRule.catLevelTwo -> mutable.Set())
       sinkSubCategories(sinkRule.catLevelTwo).add(sinkRule.nodeType.toString)
     })
 
-    sinkSubCategories.foreach(sinkSubTypeEntry =>
-      filterFlowsBySubCategoryNodeTypeAndStoreInCache(
+    val expendedSourceSinkInfo = new ListBuffer[ExpendedSourceSinkInfo]()
+
+    sinkSubCategories.foreach(sinkSubTypeEntry => {
+      filterFlowsBySubCategoryNodeType(
         dataflowMapByPathId,
         sinkSubTypeEntry._1,
         sinkSubTypeEntry._2.toSet,
         privadoScanConfig,
-        ruleCache
+        expendedSourceSinkInfo,
+        isStoreMatricData
       )
-    )
+    })
+
+    expendedSourceSinkInfo.toList
   }
 
-  /** Helper function to filter flows by sub-category and node type and store in cache
+  /** Helper function to filter flows by sub-category and node type and store
+    *
     * @param dataflowMapByPathId
     *   \- map containing pathId -> path
     * @param sinkSubCategory
     *   \- sinkSubCategory - Ex - leakages, third-parties
     * @param sinkNodetypes
     *   \- REGULAR, API etc
     */
-  private def filterFlowsBySubCategoryNodeTypeAndStoreInCache(
+
+  private def filterFlowsBySubCategoryNodeType(
     dataflowMapByPathId: Map[String, Path],
     sinkSubCategory: String,
     sinkNodetypes: Set[String],
     privadoScanConfig: PrivadoInput,
-    ruleCache: RuleCache
+    expendedSourceSinkInfo: ListBuffer[ExpendedSourceSinkInfo],
+    isStoreMatricData: Boolean
```

Comment:
assign a default value of true to this variable, so that the default behaviour is writing to metric

Reply:
i handled this at function level. okay sure will change it here.

Reply:
I handled at previous function `processExpendedSourceSinkData` because this is the main for expending the path which common among audit and main flow.

Reply:
ok, got it